Refactor the Zylevels1 chapter (#1018)
riscv/riscv-cheri: src/cheri/zylevels1.adoc

IMPORTANT: {not_v1_ratification_package_freeze}

{cheri_levels1_ext_name} adds an *Information Flow Control* (IFC) mechanism to CHERI.
It classifies capabilities as either *global* or *local* and adds checks to memory operations to control how capabilities can be shared.

> This extension adds a new metadata bit called the **Global (GL) flag** and two new permissions, **Store Local (SL)** and **Load Global (LG)**.

Reviewer: Not at all obligatory, but I wonder if something like this would make the choice of name clearer up front?

```suggestion
It adds one new "level" metadata bit to classify capabilities as either *global* or *local*, adds checks to memory operations to control how capabilities can be shared, and adds two new level-related permissions, **Store Local (SL)** and **Load Global (LG)**.
```

=== Added Metadata Bits (Permissions and Flags)

[#zylevels1_ap_field_summary,width="100%",options=header,halign=center,cols="2,2,5"]
|==============================================================================
| Bit                    | Type                   | Comment

> | <<zylevels1_gl_perm>>  | Data flow label        | Used to classify capabilities as global or local.

Reviewer: We don't define "data flow label" anywhere, AFAIR... so maybe "Information flow label" or "IFC label" or similar, here and below?

> | <<zylevels1_sl_perm>>  | Data memory permission | Used to filter the validity of stored capabilities.

Reviewer: ```suggestion
| <<zylevels1_sl_perm>>  | Data memory permission | Used to filter the {ctag}s of stored capabilities.
```

| <<zylevels1_lg_perm>>  | Data memory permission | Used to filter the permissions of loaded capabilities.
|==============================================================================

ifdef::cheri_v9_annotations[]
NOTE: The <<zylevels1_gl_perm>> and <<zylevels1_sl_perm>> are as in CHERI v9, Morello, and CHERIoT v1.
The <<zylevels1_lg_perm>> is as in CHERIoT v1.
endif::[]

[#zylevels1_gl_perm,reftext="GL(obal) Flag"]
==== The Capability Global (GL) Flag

The _Capability Global_ (GL) flag is a metadata bit used to track and control the propagation of capabilities.
It works in conjunction with the <<zylevels1_lg_perm>> and <<zylevels1_sl_perm>> permissions to enforce information flow boundaries.

The _Capability Global_ flag holds one of two values:

* 1: the capability is _global_.
* 0: the capability is _local_.

> Like permissions, the _Capability Global_ flag can be cleared when creating a new capability from an existing one, but it can never be set unless derived from a global superset capability.

Reviewer 2: ```suggestion
Like permissions, the _Capability Global_ flag can be cleared when creating a new capability from an existing one, but it can never be set unless derived from a superset capability which has the  _Capability Global_ flag set .
```

I also think we need words explaining subset / superset somewhere in the spec (see #1012 ) but doesn't need to be here.

|==============================================================================
   3+|Auth cap field         2+| Data cap field |
    h|*R*   h|*C*    h|*LG*   h| Tag h| Sealed h| Action
.3+.^|1 .3+.^| 1 .2+.^| 0 .2+.^| 1    | Yes     | Load data capability with its <<zylevels1_gl_perm>> cleared
                                      | No      | Load data capability with both its <<zylevels1_gl_perm>> and <<zylevels1_lg_perm>> cleared
                    3+| All other cases         | Load data capability with both its <<zylevels1_gl_perm>> and <<zylevels1_lg_perm>> unmodified
|==============================================================================

NOTE: <<zylevels1_lg_perm>> is only effective when both <<r_perm>> and <<c_perm>> are granted.

NOTE: Because <<zylevels1_sl_perm>> is only relevant to capabilities granting <<w_perm>>, any load that clears <<w_perm>> (e.g., due to a lack of <<lm_perm>> permission) will also necessarily clear <<zylevels1_sl_perm>>.

=== Interaction with <<STORE_CAP>>

{cheri_levels1_ext_name} adds a constraint to <<STORE_CAP>> (`{store_cap_name_lc} {cs2}, offset({cs1})`) and similar instructions.

> The capability is stored with a set {ctag} only if:

Reviewer 2: ```suggestion
The capability is stored with a {ctag} set to one only if:
```

* The capability being stored (`{cs2}`) is global (<<zylevels1_gl_perm>>=1), OR
* The authority used for the store (`{cs1}`) grants <<zylevels1_sl_perm>>.

> Otherwise, the capability is stored with a zero {ctag}.

Reviewer 2: ```suggestion
Otherwise, the capability is stored with the {ctag} set to zero.
```

> NOTE: While <<zylevels1_lg_perm>> restricts flow by reducing permissions (labeling it as local and stripping permissions), <<zylevels1_sl_perm>> restricts flow by invalidating capabilities (clearing the {ctag}).

Reviewer 2: ```suggestion
NOTE: While a lack of<<zylevels1_lg_perm>> restricts flow by reducing permissions (labeling the loaded capability as local and stripping permissions), <<zylevels1_sl_perm>> restricts flow by invalidating capabilities (by setting the {ctag} to zero).
```

.<<zylevels1_sl_perm>> effects for stored capabilities
[#zylevels1_store_summary,width="100%",options=header,halign=center,cols="1,1,1,1,5"]
|==============================================================================
   3+|Auth cap field       | Data cap field |
    h|*W*    h|*C* h|*SL* h|*GL*           h| Notes
.3+.^|1  .3+.^| 1   | 1    | X              | Store data capability unmodified
               .2+.^| 0    | 1              | Store data capability unmodified

>                            | 0              | Store data capability with {ctag} cleared

Reviewer 2: ```suggestion
                           | 0              | Store data capability with {ctag} set to zero
```

|==============================================================================

NOTE: <<zylevels1_sl_perm>> is only effective when both <<w_perm>> and <<c_perm>> are granted.

=== Interaction with <<SCSS>>

In implementations of {cheri_levels1_ext_name}, the <<SCSS>> instruction (`{SCSS_LC} rd, {cs1}, {cs2}`) indicates that `{cs2}` is a subset of `{cs1}` (setting `rd` to 1) only if:

* The Global flag of the supposed subset capability (`{cs2}`) is clear (0), OR
* The Global flag of the supposed superset capability (`{cs1}`) is set (1).

The existing permission subset logic applies to the new <<zylevels1_sl_perm>> and <<zylevels1_lg_perm>> permissions.

=== Interaction with <<CBLD>>

A <<CBLD>> instruction (`{CBLD_LC} {cd}, {cs1}, {cs2}`) may yield a valid capability (`{cd}.tag=1`) only if:

* The Global flag of the bits to be reconstituted (`{cs2}`) is clear (0), OR
* The Global flag of the reconstituting superset capability (`{cs1}`) is set (1).

The existing permission subset logic applies to the new <<zylevels1_sl_perm>> and <<zylevels1_lg_perm>> permissions.

=== Interaction with <<YSUNSEAL>>

> The <<YSUNSEAL>> instruction (`{YSUNSEAL_LC} {cd}, {cs1}, {cs2}`) ensures that the result grants <<zylevels1_gl_perm>> only if the unsealing authority (`{cs1}`) also grants <<zylevels1_gl_perm>>.

Reviewer 2: I don't think `grants` is the right verb here.
How about has Gl set?

> The <<YSUNSEAL>> instruction (`{YSUNSEAL_LC} {cd}, {cs1}, {cs2}`) ensures that the result grants <<zylevels1_gl_perm>> only if the unsealing authority (`{cs1}`) also grants <<zylevels1_gl_perm>>.

Reviewer 2: ```suggestion
The <<YSUNSEAL>> instruction (`{YSUNSEAL_LC} {cd}, {cs1}, {cs2}`) ensures that the result has<<zylevels1_gl_perm>> set only if the unsealing authority (`{cs1}`) also has <<zylevels1_gl_perm>> set.
```

                                      | No      | Load data capability with both its <<zylevels1_gl_perm>> and <<zylevels1_lg_perm>> cleared
                    3+| All other cases         | Load data capability with both its <<zylevels1_gl_perm>> and <<zylevels1_lg_perm>> unmodified
|==============================================================================

NOTE: <<zylevels1_lg_perm>> is only effective when both <<r_perm>> and <<c_perm>> are granted.

NOTE: Because <<zylevels1_sl_perm>> is only relevant to capabilities granting <<w_perm>>, any load that clears <<w_perm>> (e.g., due to a lack of <<lm_perm>> permission) will also necessarily clear <<zylevels1_sl_perm>>.

=== Interaction with <<STORE_CAP>>

{cheri_levels1_ext_name} adds a constraint to <<STORE_CAP>> (`{store_cap_name_lc} {cs2}, offset({cs1})`) and similar instructions.
The capability is stored with a set {ctag} only if:

* The capability being stored (`{cs2}`) is global (<<zylevels1_gl_perm>>=1), OR
* The authority used for the store (`{cs1}`) grants <<zylevels1_sl_perm>>.

Otherwise, the capability is stored with a zero {ctag}.

> NOTE: While <<zylevels1_lg_perm>> restricts flow by clearing metadata bits, <<zylevels1_sl_perm>> restricts flow by clearing the {ctag}.

Reviewer: Maybe emphasize the semantic, not syntactic, change?
```suggestion
NOTE: While <<zylevels1_lg_perm>> restricts flow by attenuation of a valid capability (labeling it as local and stripping permissions), <<zylevels1_sl_perm>> restricts flow by invalidating capabilities (clearing the {ctag}).
```

Author: Thanks that's a good suggestion. I'll probably change it a tiny bit to avoid the word attenuation since that's not commonly used vocabulary.

.<<zylevels1_sl_perm>> effects for stored capabilities
[#zylevels1_store_summary,width="100%",options=header,halign=center,cols="1,1,1,1,5"]
|==============================================================================
   3+|Auth cap field       | Data cap field |
    h|*W*    h|*C* h|*SL* h|*GL*           h| Notes
.3+.^|1  .3+.^| 1   | 1    | X              | Store data capability unmodified
               .2+.^| 0    | 1              | Store data capability unmodified
                           | 0              | Store data capability with {ctag} cleared
|==============================================================================

NOTE: <<zylevels1_sl_perm>> is only effective when both <<w_perm>> and <<c_perm>> are granted.

=== Interaction with <<SCSS>>

In implementations of {cheri_levels1_ext_name}, the <<SCSS>> instruction (`{SCSS_LC} rd, {cs1}, {cs2}`) indicates that `{cs2}` is a subset of `{cs1}` (setting `rd` to 1) only if:

* The Global flag of `{cs2}` is clear (0), OR

> * The Global flag of `{cs1}` is set (1).

Reviewer: ```suggestion
* The Global flag of the supposed subset capability (`{cs2}`) is clear (0), OR
* The Global flag of the supposed superset capability (`{cs1}`) is set (1).
```

The existing permission subset logic applies to the new <<zylevels1_sl_perm>> and <<zylevels1_lg_perm>> permissions.

=== Interaction with <<CBLD>>

A <<CBLD>> instruction (`{CBLD_LC} {cd}, {cs1}, {cs2}`) may yield a valid capability (`{cd}.tag=1`) only if:

* The Global flag of `{cs2}` is clear (0), OR

> * The Global flag of `{cs1}` is set (1).

Reviewer: ```suggestion
* The Global flag of the bits to be reconstituted (`{cs2}`) is clear (0), OR
* The Global flag of the reconstituting superset capability (`{cs1}`) is set (1).
```
